File: fsharp/yacht/Yacht.fs
module Yacht

type Category =
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One = 1
    | Two = 2
    | Three = 3
    | Four = 4
    | Five = 5
    | Six = 6

let score category dice =
    let straight =
        [ Die.One
          Die.Two
          Die.Three
          Die.Four
          Die.Five
          Die.Six ]

    let isLittleStraight dice =
        List.sortBy int dice = List.take 5 straight

    let isBigStraight dice =
        List.sortBy int dice = List.skip 1 straight

    let isFullHouse dice =
        let counts = dice |> Seq.countBy id |> Seq.map snd
        Seq.length counts = 2 && Seq.exists ((=) 3) counts

    let isYacht dice = dice |> Seq.distinct |> Seq.length = 1

    let scoreNumbers dice die =
        (int die)
        * (dice |> Seq.filter ((=) die) |> Seq.length)

    let scoreFourOfAKind =
        Seq.countBy id
        >> Seq.filter (fun (_, count) -> count >= 4)
        >> Seq.map fst
        >> Seq.sumBy (fun die -> (int die) * 4)

    let sumOfDice = Seq.sumBy int

    match category with
    | Ones -> scoreNumbers dice Die.One
    | Twos -> scoreNumbers dice Die.Two
    | Threes -> scoreNumbers dice Die.Three
    | Fours -> scoreNumbers dice Die.Four
    | Fives -> scoreNumbers dice Die.Five
    | Sixes -> scoreNumbers dice Die.Six
    | Choice -> sumOfDice dice
    | FourOfAKind -> scoreFourOfAKind dice
    | FullHouse when isFullHouse dice -> sumOfDice dice
    | LittleStraight when isLittleStraight dice -> 30
    | BigStraight when isBigStraight dice -> 30
    | Yacht when isYacht dice -> 50
    | _ -> 0

File: kotlin/yacht/src/main/kotlin/Yacht.kt
import YachtCategory.*

object Yacht {

    fun solve(category: YachtCategory, vararg dices: Int) = when (category) {
        ONES, TWOS, THREES, FOURS, FIVES, SIXES -> dices.filter { it == category.ordinal }.sum()
        YACHT -> scoreWhen(dices.isYacht, 50)
        FULL_HOUSE -> scoreWhen(dices.isFullHouse, dices.sum())
        FOUR_OF_A_KIND -> dices.scoreFourOfAKind()
        LITTLE_STRAIGHT -> scoreWhen(dices.isLittleStraight, 30)
        BIG_STRAIGHT -> scoreWhen(dices.isBigStraight, 30)
        CHOICE -> dices.sum()
    }

    private fun scoreWhen(condition: Boolean, score: Int) = if (condition) score else 0

    private val IntArray.isYacht get() = all { it == first() }

    private val IntArray.isFullHouse get() = groupBy { it }.values.map { it.size }.toSet() == setOf(2, 3)

    private val IntArray.isLittleStraight get() = sorted() == (1..5).toList()

    private val IntArray.isBigStraight get() = sorted() == (2..6).toList()

    private fun IntArray.scoreFourOfAKind() =
        groupBy { it }
            .values
            .firstOrNull { it.size >= 4 }
            ?.take(4)
            ?.sum()
            ?: 0
}
